Sources/PerformanceTesting/DataSet.swift

```diff
@@ -0,0 +1,121 @@
+//
+//  DataSet.swift
+//  PerformanceTesting
+//
+//  Created by James Bean on 8/10/17.
+//
+
+import Darwin
+
+/// Collection of (x,y) data points.
+struct DataSet {
+
+    struct LinearRegression {
+        let slope: Double
+        let intercept: Double
+        let correlation: Double
+    }
+
+    // MARK: - Instance Properties
+
+    /// - Returns: Linear regression of data contained herein.
+    var linearRegression: LinearRegression {
+
+        let (xs,ys) = unzip(data)
+        let sumOfXs = xs.sum
+        let sumOfYs = ys.sum
+        let sumOfXsSquared = xs.map { pow($0,2) }.sum
+        let sumOfXsTimesYs = data.map(*).sum
+
+        let denominator = Double(data.count) * sumOfXsSquared - pow(sumOfXs, 2)
+        let interceptNumerator = sumOfYs * sumOfXsSquared - sumOfXs * sumOfXsTimesYs
+        let slopeNumerator = Double(data.count) * sumOfXsTimesYs - sumOfXs * sumOfYs
+        let intercept = interceptNumerator / denominator
+        let slope = slopeNumerator / denominator
+
+        return LinearRegression(
+            slope: slope,
+            intercept: intercept,
+            correlation: correlation(data, sumOfXs: sumOfXs, sumOfYs: sumOfYs, slope: slope)
+        )
+    }
+
+    let data: [(Double,Double)]
+
+    // MARK: - Initializers
+
+    init(_ data: [(Double,Double)]) {
+        self.data = data
+    }
+
+    // MARK: - Instance Methods
+
+    /// Maps data representing performance of a certain complexity so that it
+    /// can be fit with linear regression. This is done by applying the inverse
+    /// function of the expected performance function.
+    func mappedForLinearFit(complexity: Complexity) -> DataSet {
+        return DataSet(data.map { ($0, complexity.inverse($1)) })
+    }
+
+    /// - Returns: `true` if the curve of this data resembles the given `complexity` curve.
+    /// Otherwise, returns `false`.
+    func curve(
+        is complexity: Complexity,
+        tolerance: Double = 0.1,
+        minimumCorrelation: Double = 0.9
+    ) -> Bool
+    {
+        switch complexity {
+        case .constant:
+            let results = linearRegression
+            return (
+                approximatelyEqual(results.slope, 0, epsilon: tolerance) ||
```

Comment:
I was getting some strange behavior (failing) with `.constant` time accesses. I changed this from `&&` to `||` to see if they were actually testing against different cases.

@brianlheim, could you explain a little bit how the `.correlation` condition works here?

Reply:
this was an experimental tweak; IIRC as the regression slope approaches constant it becomes unstable (or more sensitive to variance?). the second condition is added because from an intuitive perspective if time is not strongly correlated with input size it's probably not actually dependent (i.e. constant complexity). totally a hack, maybe we should scale the tolerance in this case

Reply:
Oh neat. Do you think it would be worth it to bump up the `trialCount` for `.constant` queries in order to smooth out the data, or even repeating the accesses `100` or `10_000` within a given trial?

Reply:
The tests seem to be humming along nicely, without the `.correlation` check, and with the `tolerance` at `0.5`.

Reply:
.5 should be fine i think? do we have a test to catch cases of linear time masquerading as constant?

Reply:
Ope, whoops, and it should be: `0.05`.

We don't, but I was thinking of doing something like that to whittle down what the tolerances should be.